From stable-bounces@linux.kernel.org Wed Oct 25 23:51:42 2006
Date: Wed, 25 Oct 2006 23:51:07 -0700 (PDT)
Message-Id: <20061025.235107.39158283.davem@davemloft.net>
To: stable@kernel.org
From: David Miller <davem@davemloft.net>
Mime-Version: 1.0
Subject: SPARC64: Fix memory corruption in pci_4u_free_consistent().
Content-Type: text/plain; charset="us-ascii"

From: David Miller <davem@davemloft.net>

The second argument to free_npages() was being incorrectly
calculated, which would thus access far past the end of the
arena->map[] bitmap.

Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
Signed-off-by: Chris Wright <chrisw@sous-sol.org>

---
 arch/sparc64/kernel/pci_iommu.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- linux-2.6.18.1.orig/arch/sparc64/kernel/pci_iommu.c
+++ linux-2.6.18.1/arch/sparc64/kernel/pci_iommu.c
@@ -281,7 +281,7 @@ static void pci_4u_free_consistent(struc
 
 	spin_lock_irqsave(&iommu->lock, flags);
 
-	free_npages(iommu, dvma, npages);
+	free_npages(iommu, dvma - iommu->page_table_map_base, npages);
 
 	spin_unlock_irqrestore(&iommu->lock, flags);
 
